### PR TITLE
feat: be able to create scatter plots

### PR DESCRIPTION
**Pull request by Marc-AntoineA**
### What
Enrich the API to not only display bar charts but also scatter plots.

### Screenshot
image

### CODE REVIEW

File: app/api.py [hunk 1]
```diff
@@ -107,14 +103,13 @@ def search_get(
     fields: GetSearchParamsTypes.fields = None,
     sort_by: GetSearchParamsTypes.sort_by = None,
     facets: GetSearchParamsTypes.facets = None,
-    charts: GetSearchParamsTypes.charts = None,
```
Comment (alexgarel) [inline app/api.py:110]: I was thinking it was good to keep your field_name,x:y pattern here (and transform it inside the method), it offers a simple interface for simple use cases.

File: frontend/src/search-chart.ts [hunk 2]
```diff
@@ -20,85 +19,54 @@ export class SearchaliciousChart extends SearchaliciousResultCtlMixin(
...
-    return html`<div id="${this.name!}"></div>`;
-  }
-
-  override render() {
-    return html` <div class="white-panel">${this.renderChart()}</div>`;
-  }
-
+  // Vega function assumes that rendered had been previously
+  // called.
   override handleResults(event: SearchResultEvent) {
```
Comment (alexgarel) [inline frontend/src/search-chart.ts:28]: why isn't that in the mixin ? It seems to be the same for both charts.

File: frontend/src/mixins/search-ctl.ts [hunk 3]
```diff
@@ -273,9 +262,33 @@ export const SearchaliciousSearchMixin = <T extends Constructor<LitElement>>(
...
+        field: node.name,
+      }));
+
+      const scatterCharts = Array.from(
+        document.querySelectorAll(
+          `searchalicious-scatter-chart[search-name=${this.name}]`
+        )
+      ).map((node) => ({
+        chart_type: 'ScatterChartType',
+        // @ts-ignore
```
Comment (alexgarel) [inline frontend/src/mixins/search-ctl.ts:282]: I would prefer that we specify the ts rule we ignore. Also to make sense of what we need to ignore.

**Review by alexgarel — CHANGES_REQUESTED** [review]
@Marc-AntoineA I made some comments.

Also an important thing to change: add comments, at least to every class and to big blocks of codes. At least a summary of what you are doing. (and for classes the big picture).

File: frontend/src/mixins/search-ctl.ts (hunk 3 above)
Comment (alexgarel) [reply frontend/src/mixins/search-ctl.ts:282]: This also applies to the other `// @ts-ignore`.

Thanks in advance :-)

File: app/_types.py [hunk 4]
```diff
@@ -323,7 +341,18 @@ def check_facets_are_valid(self):
...
+
+        errors = check_all_values_are_fields_agg(
+            self.index_id,
+            [
+                chart.field
+                for chart in self.charts
+                if chart.chart_type == "DistributionChart"
+            ],
+        )
+        # TODO check for x and y fields?
```
[the remark this answers is not in the document]
Comment (Marc-AntoineA) [reply app/_types.py:355]: In fact some fields are not numeric but they are subfields of objects fields. Ex. `nutriments.fiber_100g` :/

File: app/api.py (hunk 1 above)
Comment (Marc-AntoineA) [reply app/api.py:110]: Done.

File: app/charts.py [hunk 5]
```diff
@@ -1,122 +1,234 @@
...
+                    ],
                 },
-            ],
+            },
+        },
+    ]
+    return chart
+
+
+def build_scatter_chart(chart_option: ScatterChartType, search_result):
```
[the remark this answers is not in the document]
Comment (Marc-AntoineA) [reply app/charts.py:123]: Oh yes, it’s only for the current page. Nice spot.

File: app/charts.py [hunk 6]
```diff
@@ -1,122 +1,234 @@
...
+            "from": {"data": "source"},
+            "encode": {
+                "update": {
+                    "x": {"scale": "x", "field": vega_x},
+                    "y": {"scale": "y", "field": vega_y},
+                    "shape": {"value": "circle"},
+                    "strokeWidth": {"value": 2},
+                    "opacity": {"value": 0.5},
+                    "stroke": {"value": PRIMARY_COLOR},
+                    "fill": {"value": "transparent"},
```
[the remark this answers is not in the document]
Comment (Marc-AntoineA) [reply app/charts.py:192]: If you want, it was the default value from the example.

File: frontend/src/mixins/search-chart.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  // eslint-disable-next-line
+  vegaRepresentation: any;
+
+  testVegaInstalled(): void;
+  getName(): string;
+}
+
+export const SearchaliciousChartMixin = <T extends Constructor<LitElement>>(
+  superClass: T
+) => {
```
[the remark this answers is not in the document]
Comment (Marc-AntoineA) [reply frontend/src/mixins/search-chart.ts:31]: Changed to simple class with inheritance

File: app/_types.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Marc-AntoineA) [reply app/_types.py]: Checked.

File: frontend/src/mixins/search-ctl.ts (hunk 3 above)
Comment (Marc-AntoineA) [reply frontend/src/mixins/search-ctl.ts]: Removed

File: app/validations.py [hunk 8]
```diff
@@ -37,7 +37,48 @@ def check_all_values_are_fields_agg(
...
+    return errors
+
+
+def check_fields_are_numeric(
+    index_id: str | None, values: list[str] | None
+) -> list[str]:
+    """
+    * If field exists in global_config, check that it's numeric.
+    * If field looks like x.y and x.y does not exist in global config,
+    check that x is a n object field
```
Comment (alexgarel) [inline app/validations.py:52]:
```suggestion
    check that x is an object field
```

**Review by alexgarel — APPROVED** [review]
@Marc-AntoineA this is almost ok ! (just one mandatory thing to fix: the typing in search_get).

Thanks for this PR.

**Pull request merged**